### ISSUE TITLE
Arduino IDE 2.x (for ARM64)

### ISSUE THREAD

**Issue by matu6968** [Botspot/pi-apps#2585]
### What is the name of the app?

Arduino IDE 2.x

### Where is the app hosted?

https://github.com/arduino/arduino-ide

### About the app

The newer and redefined Arduino IDE which allows you to code for embedded platforms (like the older 1.8.x based IDE, such as for Arduino Uno), supports many extensions just like the 1.8.x based one.

### Upload file or Add PR Link

[arduino-ide-2.x.zip](https://github.com/Botspot/pi-apps/files/15140080/arduino-ide-2.x.zip)

### Confirmations

- [X] I have confirmed that this app is legal and not piracy.
- [X] I have confirmed that this app has never been discussed in https://github.com/Botspot/pi-apps/issues and https://github.com/Botspot/pi-apps/pulls, and it is not in the Pi-Apps app list.
- [X] I have read and followed the [creating an app for Pi-Apps documentation](https://pi-apps.io/wiki/development/Creating-an-app/).
- [X] I have confirmed that this app can run on all [Pi-Apps supported systems](https://pi-apps.io/install/).

**Comment by Botspot** [comment]
Who owns this domain? https://ximitekcloud.com/downloads
Also this seems to be an unofficial build, so can we be sure that it will continue to receive updates in the future?

**Comment by matu6968** [comment]
> Who owns this domain? https://ximitekcloud.com/downloads Also this seems to be an unofficial build, so can we be sure that it will continue to receive updates in the future?

I own it, as its a unofficial build i will also make updates for it as its kinda odd for Arduino to make a ARM64 version for Mac while ignoring Linux.

**Comment by Botspot** [comment]
> I own it, as its a unofficial build i will also make updates for it as its kinda odd for Arduino to make a ARM64 version for Mac while ignoring Linux.

OK, so why not then host the build scripts and release assets on a github repo?

**Comment by matu6968** [comment]
> > I own it, as its a unofficial build i will also make updates for it as its kinda odd for Arduino to make a ARM64 version for Mac while ignoring Linux.
> 
> OK, so why not then host the build scripts and release assets on a github repo?

I couldn't host the release assets on GitHub as when i tried a git pushing the repo with the main release assets, it just refused to do it (i was doing it on a private repo, will try on a public one  and will reply to see if it does something)

**Comment by Botspot** [comment]
> I couldn't host the release assets on GitHub as when i tried a git pushing the repo with the main release assets, it just refused to do it (i was doing it on a private repo, will try on a public one and will reply to see if it does something)

If builds will only be periodic, consider the option to simply upload them manually.

**Comment by matu6968** [comment]
> > I couldn't host the release assets on GitHub as when i tried a git pushing the repo with the main release assets, it just refused to do it (i was doing it on a private repo, will try on a public one and will reply to see if it does something)
> 
> If builds will only be periodic, consider the option to simply upload them manually.

As my Pi 5 (the Pi i used for compiling the IDE) isn't used heavy duty by another project, i could do like minor release versions when they come out (for example 2.3.3 > 2.3.4)

**Comment by Botspot** [comment]
> > Who owns this domain? https://ximitekcloud.com/downloads
> 
> I own it

What are the chances of it going down? Is it self hosted?

**Comment by matu6968** [comment]
> > > Who owns this domain? https://ximitekcloud.com/downloads
> > 
> > 
> > I own it
> 
> What are the chances of it going down? Is it self hosted?

It is just a Linux based web server hosted on Linode with Caddy installed as the web server.

**Comment by matu6968** [comment]
> > I own it, as its a unofficial build i will also make updates for it as its kinda odd for Arduino to make a ARM64 version for Mac while ignoring Linux.
> 
> OK, so why not then host the build scripts and release assets on a github repo?

And still nope, i cannot host the release assets on GitHub even when the repo is public.

**Comment by theofficialgman** [comment]
I am going to close this. Not interested in another porting project and still waiting on Arduino to show that they care to maintain their IDE project. There has not been a commit since February and they are running an old version of electron with security issues.

This has already been requested of upstream by many many users. It is trivial for them to implement. Please add your request to the others there. If you are able to help make any modifications to their build CI to build arm64 linux releases then by all means help them out.

**Comment by matu6968** [comment]
> I am going to close this. Not interested in another porting project and still waiting on Arduino to show that they care to maintain their IDE project. There has not been a commit since February and they are running an old version of electron with security issues.
> 
> This has already been requested of upstream by many many users. It is trivial for them to implement. Please add your request to the others there. If you are able to help make any modifications to their build CI to build arm64 linux releases then by all means help them out with a PR.

I just recently moved the package to a [GitHub repo](https://github.com/matu6968/arduino-ide-arm64) and provided some instructions to build it so no more need for a external web server anymore. Also as @theofficialgman does not like the current state of the Arduino IDE 2.x (due to using a outdated Electron 27 build) then i do not think it will be ever published until Arduino updates Electron to the latest version and to contribute to the issues (as well as also responding to backlash on why they didn't ship a Linux ARM64 version of their IDE officially).

**Comment by Botspot** [comment]
This should be revisited, now with more regular updates and their current dependence on `"node": ">=18.17.0 <21"`

**Comment by Botspot** [comment]
The build from @matu6968 seems to be offline.
Try again hosting release assets on github, or provide a script to convert the linux release to arm64.

**Comment by theofficialgman** [comment]
> @theofficialgman does not like the current state of the Arduino IDE 2.x (due to using a outdated Electron 27 build) then i do not think it will be ever published until Arduino updates Electron to the latest version and to contribute to the issues (as well as also responding to backlash on why they didn't ship a Linux ARM64 version of their IDE officially).

This issue still exists... There are so many critical bugs that have been resolved since 27 in August 2023, not to mention all the security vulnerabilities that come with such an out of date and unmaintained version of chromium. I simply will not accept new applications that use electron and don't do the basic maintenance to keep up to date with one of the stable electron releases (major version releases are supported for 8 months, they don't even need update frequently).

In any case, our current stance on such porting projects is for them to not exist when an effort could be contributed upstream. This seems like a prime candidate and would resolve an issue for many many users who have requested such a feature https://github.com/arduino/arduino-ide/issues/107 . Consider using your time to create CI for upstream to build for arm64/armhf linux.

**Comment by matu6968** [comment]
> The build from @matu6968 seems to be offline.

Regarding about the hosting before it got migrated my github repo, the server for that is long gone. 

> Try again hosting release assets on github, or provide a script to convert the linux release to arm64.

Since Arduino released their official 2.3.3 version of their 2.x based IDE, i will possibly today make a CI and/or a bash script to automate the build process and not to worry about updates if I made a CI (as long as Arduino themselves don't make breaking changes to their code especially with changing Electron versions to a newer version)

**Comment by matu6968** [comment]
> The build from @matu6968 seems to be offline. Try again hosting release assets on github, or provide a script to convert the linux release to arm64.

The script is ready, except for checking if golang is installed (i will add the golang installed check later and put it on my repo)
[build-ide.txt](https://github.com/user-attachments/files/17758848/build-ide.txt)
(rename build-ide.txt to build-ide.sh and enable execution permissions to make it work)

**Comment by Botspot** [comment]
Did you ever encounter this error @matu6968?
```
pi@raspberrypi:~/arduino-ide $ /home/pi/Downloads/build-ide.sh build-package
64 bit ARM Architecture detected
v22.11.0
Now using node v20.18.0 (npm v10.8.2)
building package
installing node dependencies
yarn install v1.22.22
[1/5] Validating package.json...
[2/5] Resolving packages...
[3/5] Fetching packages...
warning vscode-arduino-api@0.1.2: The engine "vscode" appears to be invalid.
[4/5] Linking dependencies...
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > @theia/electron@1.41.0" has unmet peer dependency "electron@^23.2.4".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > @tippyjs/react@4.2.6" has unmet peer dependency "react@>=16.8".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > @tippyjs/react@4.2.6" has unmet peer dependency "react-dom@>=16.8".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-disable@0.1.1" has unmet peer dependency "react@>=16".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-markdown@8.0.7" has unmet peer dependency "@types/react@>=16".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-markdown@8.0.7" has unmet peer dependency "react@>=16".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-perfect-scrollbar@1.5.8" has unmet peer dependency "react@>=16.3.3".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-perfect-scrollbar@1.5.8" has unmet peer dependency "react-dom@>=16.3.3".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-select@5.7.7" has unmet peer dependency "react@^16.8.0 || ^17.0.0 || ^18.0.0".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-select@5.7.7" has unmet peer dependency "react-dom@^16.8.0 || ^17.0.0 || ^18.0.0".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-tabs@3.2.3" has unmet peer dependency "react@^16.3.0 || ^17.0.0-0".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-window@1.8.9" has unmet peer dependency "react@^15.0.0 || ^16.0.0 || ^17.0.0 || ^18.0.0".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-window@1.8.9" has unmet peer dependency "react-dom@^15.0.0 || ^16.0.0 || ^17.0.0 || ^18.0.0".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > @theia/scm > react-autosize-textarea@7.1.0" has unmet peer dependency "react@^0.14.0 || ^15.0.0 || ^16.0.0".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > @theia/scm > react-autosize-textarea@7.1.0" has unmet peer dependency "react-dom@^0.14.0 || ^15.0.0 || ^16.0.0".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-select > @emotion/react@11.11.1" has unmet peer dependency "react@>=16.8.0".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-select > react-transition-group@4.4.5" has unmet peer dependency "react@>=16.6.0".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-select > react-transition-group@4.4.5" has unmet peer dependency "react-dom@>=16.6.0".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-select > use-isomorphic-layout-effect@1.1.2" has unmet peer dependency "react@^16.8.0 || ^17.0.0 || ^18.0.0".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > @theia/debug > @theia/process > node-pty@0.11.0-beta17" has incorrect peer dependency "node-gyp@^8.3.0".
warning "workspace-aggregator-92977e9f-fac2-4112-9b6a-a209b636165a > arduino-ide-extension > react-select > @emotion/react > @emotion/use-insertion-effect-with-fallbacks@1.0.1" has unmet peer dependency "react@>=16.8.0".
[5/5] Building fresh packages...
$ lerna run prepare
lerna notice cli v7.4.2
lerna info versioning independent

 >  Lerna (powered by Nx)   Running target prepare for 2 projects:

    - arduino-ide-extension
    - electron-app

 ————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

> arduino-ide-extension:prepare

arduino-ide-extension: yarn run v1.22.22
arduino-ide-extension: $ yarn download-cli && yarn download-fwuploader && yarn download-ls && yarn copy-i18n && yarn download-examples
$ node ./scripts/download-cli.js
arduino-ide-extension: 📦  Identified released version of the CLI. Downloading version 1.0.4 from 'https://downloads.arduino.cc/arduino-cli/arduino-cli_1.0.4_Linux_ARM64.tar.gz'
arduino-ide-extension: >>> Downloading from 'https://downloads.arduino.cc/arduino-cli/arduino-cli_1.0.4_Linux_ARM64.tar.gz'...
arduino-ide-extension: <<< Download succeeded.
arduino-ide-extension: >>> Decompressing...
arduino-ide-extension: <<< Decompressing succeeded.
arduino-ide-extension: Done: /home/pi/arduino-ide/arduino-ide-extension/src/node/resources/arduino-cli
arduino-ide-extension: $ node ./scripts/download-fwuploader.js
arduino-ide-extension: 📦  Identified released version of the Firmware Uploader. Downloading version 2.4.1 from 'https://downloads.arduino.cc/arduino-fwuploader/arduino-fwuploader_2.4.1_Linux_ARM64.tar.gz'
arduino-ide-extension: >>> Downloading from 'https://downloads.arduino.cc/arduino-fwuploader/arduino-fwuploader_2.4.1_Linux_ARM64.tar.gz'...
arduino-ide-extension: <<< Download succeeded.
arduino-ide-extension: >>> Decompressing...
arduino-ide-extension: <<< Decompressing succeeded.
arduino-ide-extension: Done: /home/pi/arduino-ide/arduino-ide-extension/src/node/resources/arduino-fwuploader
arduino-ide-extension: $ node ./scripts/download-ls.js
arduino-ide-extension: Building language-server from https://github.com/arduino/arduino-language-server.git. Commitish: 05ec308
arduino-ide-extension: >>> Cloning language-server source to /tmp/d-20241014-972600-2ai2tu.h1aev...
arduino-ide-extension: Cloning into '/tmp/d-20241014-972600-2ai2tu.h1aev'...
arduino-ide-extension: <<< Cloned language-server repo.
arduino-ide-extension: >>> Checking out 05ec308...
arduino-ide-extension: Note: switching to '05ec308'.
arduino-ide-extension: You are in 'detached HEAD' state. You can look around, make experimental
arduino-ide-extension: changes and commit them, and you can discard any commits you make in this
arduino-ide-extension: state without impacting any branches by switching back to a branch.
arduino-ide-extension: If you want to create a new branch to retain commits you create, you may
arduino-ide-extension: do so (now or later) by using -c with the switch command. Example:
arduino-ide-extension:   git switch -c <new-branch-name>
arduino-ide-extension: Or undo this operation with:
arduino-ide-extension:   git switch -
arduino-ide-extension: Turn off this advice by setting config variable advice.detachedHead to false
arduino-ide-extension: HEAD is now at 05ec308 Merge pull request #191 from arduino/update_grpc
arduino-ide-extension: <<< Checked out 05ec308.
arduino-ide-extension: 05ec308
arduino-ide-extension: >>> Building the language-server...
arduino-ide-extension: Failed to execute go with args: ["build"]
arduino-ide-extension: Error: spawnSync go ENOENT
arduino-ide-extension: error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
arduino-ide-extension: error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.

 ————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————————

 >  Lerna (powered by Nx)   Running target prepare for 2 projects failed

   Tasks not run because their dependencies failed or --nx-bail=true:
   
   - electron-app:prepare
   
   Failed tasks:
   
   - arduino-ide-extension:prepare

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.

!!! An error happened during script execution
!!! Please check console output for bad sync,
!!! failed patch application, OOM killer messages, etc.
```

**Comment by Botspot** [comment]
Found a solution here: https://github.com/arduino/arduino-ide/issues/2499#issuecomment-2316715572

**Comment by Botspot** [comment]
Fixed and working script for arm64 debian/ubuntu only
[build-ide.txt](https://github.com/user-attachments/files/17760751/build-ide.txt)

**Comment by matu6968** [comment]
> @matu6968 Would you be willing to try adding the arm64 arch to Arduino's official CI build? I think the only file with needed changes is this one: https://github.com/arduino/arduino-ide/blob/main/.github/workflows/build.yml#L389
> 
> The changes needed may look similar to this PR for a similar project: https://github.com/Eugeny/tabby/pull/5907/files

These changes seem to add checking environment variables for getting armhf and arm64 support, however I believe this should not be needed because you can specify electron-builder configs in package.json inside of the electron-app folder to add other arches like armhf without needing to modify the workflow.

**Comment by Botspot** [comment]
@matu6968, clearly you understand electron more than I do. Are you willing to open a pull request to arduino-ide to get arm build artifacts released alongside their x86 appimage and zip? If so, I will wait for you. If not, then I will see about doing something hacky here that theofficialgman would probably not be too fond of... ;)

**Comment by matu6968** [comment]
> @matu6968, clearly you understand electron more than I do. Are you willing to open a pull request to arduino-ide to get arm build artifacts released alongside their x86 appimage and zip? If so, I will wait for you. If not, then I will see about doing something hacky here that theofficialgman would probably not be too fond of... ;)

I will see how it turns out and make a pull request if cross compiling works, also imma see if I can update Electron 27  dependency to the latest version (which is version 27.3.11) without breaking changes.

**Comment by matu6968** [comment]
> Did you ever encounter this error @matu6968?
> ...

This is cause the script didn't check if go was installed and hence the error because go wasn't installed to build the language-server.

**Comment by Botspot** [comment]
Yeah I know, I fixed it in the script I sent back to you here. Please check it. It is simplified and only for debian arm64, but it fixed several issues you had in your script with how arduino-ide was being compiled and packaged. I think if you run your original script on the master branch of arduino-ide, you will see that much of it fails. I had to redo major parts of it based on their buildscripts.

**Comment by theofficialgman** [comment]
> @matu6968 Would you be willing to try adding the arm64 arch to Arduino's official CI build? I think the only file with needed changes is this one: https://github.com/arduino/arduino-ide/blob/main/.github/workflows/build.yml#L389 The changes needed may look similar to this PR for a similar project: https://github.com/Eugeny/tabby/pull/5907/files

To extend that, when building native libraries with yarn are involved you also need to manually specify and install the cross compilers when building on x86_64 for armhf/arm64 (eg: https://github.com/shiftkey/desktop/pull/897/files which was later refactored into a dockerfile for some other reasons https://github.com/shiftkey/desktop-ubuntu-arm64-packaging/blob/shiftkey-patch-1/Dockerfile)

**Comment by matu6968** [comment]
> To extend that, when building native libraries with yarn are involved you also need to manually specify and install the cross compilers when building on x86_64 for armhf/arm64 (eg: https://github.com/shiftkey/desktop/pull/897/files which was later refactored into a dockerfile for some other reasons https://github.com/shiftkey/desktop-ubuntu-arm64-packaging/blob/shiftkey-patch-1/Dockerfile)

I am trying to get the IDE to work on armhf but ran into a wall of problems.

Here is what I tried:

1. Cross compiling using arm64 system but with 32 bit cross compiler and some 32 bit native libraries: (note arduino-ide-extension and other arduino things are not cross compiled for armhf) After running `yarn --cwd electron-app rebuild ` , compiling errors out: https://pastebin.com/g67zgMfM 
2. Using armhf system with all the 32 bit libraries and patched out download-ls.js file to allow downloading of 32 bit clang binaries: After running `yarn --cwd electron-app rebuild ` Errors out with 2 possible outcomes: 1. a stack trace and a error with the lines: Allocation failed - JavaScript heap out of memory or 2. SIGABRT or SIGSEGV exit code with the error line for SIGSEGV: `terminate called after throwing an instance of 'std::bad_alloc'
  what():  std::bad_alloc` (meant out of memory, but 32 bit OS has a hard memory limit of around 4 GB but adding in my enviroment variables NODE_OPTIONS= --max_old_space_size=4000 did not help nor decreasing running processes in parallel using webpack config flag: `parallelism`)

**Comment by Botspot** [comment]
Reopening this issue because I still want to see Arduino 2.x added at some point. Maybe now it will be easier for arm64 with github actions supporting this architecture now.

**Comment by DubbsPi** [comment]
I am going to try to create a script to build the IDE from source.

**Comment by matu6968** [comment]
> Reopening this issue because I still want to see Arduino 2.x added at some point. Maybe now it will be easier for arm64 with github actions supporting this architecture now.

> This has already been requested of upstream by many many users. It is trivial for them to implement. Please add your request to the others there. If you are able to help make any modifications to their build CI to build arm64 linux releases then by all means help them out with a PR.

There is a open PR that adds ARM64 runners: https://github.com/arduino/arduino-ide/pull/2670

> I am going to close this. Not interested in another porting project and still waiting on Arduino to show that they care to maintain their IDE project. There has not been a commit since February and they are running an old version of electron with security issues.

They have updated Electron and the core compoments ...but a still a EOL version of Electron

https://github.com/arduino/arduino-ide/blob/main/electron-app/package.json

**Comment by neochaos42** [comment]
Hi PR author here. For now you can download the files from the build artifacts. I weekly sync my fork with Arduino main 
https://github.com/arduino/arduino-ide/pull/2670/checks

**Comment by matu6968** [comment]
> Hi PR author here. For now you can download the files from the build artifacts. I weekly sync my fork with Arduino main https://github.com/arduino/arduino-ide/pull/2670/checks

For Pi-Apps, a .deb format is better suited here since all you need to do to make the install script is a simple `install_packages` to the deb file and it is installed.

I also have my own builds hosted at https://github.com/matu6968/arduino-ide-arm64/releases/tag/v2.3.7 that have a newer version of Electron (31.7.7 is the latest that the IDE can use without build errors) and the related [.patch file](https://github.com/matu6968/arduino-ide-arm64/blob/main/add-home-page-email-deb.diff) to update dependencies (which include Electron) and make it generate .deb files by adding required fields.

**Comment by neochaos42** [comment]
> > Hi PR author here. For now you can download the files from the build artifacts. I weekly sync my fork with Arduino main https://github.com/arduino/arduino-ide/pull/2670/checks
> 
> For Pi-Apps, a .deb format is better suited here since all you need to do to make the install script is a simple `install_packages` to the deb file and it is installed.
> 
> I also have my own builds hosted at https://github.com/matu6968/arduino-ide-arm64/releases/tag/v2.3.7 that have a newer version of Electron (31.7.7 is the latest that the IDE can use without build errors) and the related [.patch file](https://github.com/matu6968/arduino-ide-arm64/blob/main/add-home-page-email-deb.diff) to update dependencies (which include Electron) and make it generate .deb files by adding required fields.

I might open a separate PR for debian packed builds.

**Comment by neochaos42** [comment]
Separate PR made https://github.com/arduino/arduino-ide/pull/2714

**Comment by neochaos42** [comment]
https://github.com/arduino/arduino-ide/pull/2739
well this might be official soon.
